### PR TITLE
Iss749 add compression

### LINKED ISSUE #454: Add option to lazy load data from object store
We could add an option to lazily load data from the object store. Large datasets such as footprints / flux aren't very big when compressed as a NetCDF. We could just return the bytes from the object store instead of doing a `load_dataset` in Datasource. We could then mayeb use an BytesIO buffer to open the data in chunks when needed.

### LINKED ISSUE #749: Add compression to footprints while waiting for zarr
### What is your issue?

Right now footprints aren't being compressed. When we switch to zarr as a storage backend, we will get compression back, but in the short term we're having trouble adding footprints to the shared object store.

We can use this work around to compress netcdf: [https://github.com/openghg/openghg_inversions/pull/19](https://github.com/openghg/openghg_inversions/pull/19).

Also, we could run a script on Blue Pebble to compress existing files once the workaround is tested.

### PR DESCRIPTION
**Pull request by brendan-m-murphy**
* **Summary of changes** (Bug fix, feature, docs update, ...)

1. Compression level for saving Datasources is increased from 4 to 5.
(Compression level 4 seems to be the default.) I added a test to show
that increasing compression does actually result in small files on disk.

2. `Datasource.load_dataset` now uses `xarray.open_dataset` instead
of `xarray.load_dataset`, which should lazy-load data. This gives a 
33% decrease in the time taken to run all of the tests on my machine.

I couldn't make a useful test to demonstrate that data is lazy-loaded,
but the xarray docs state that this is the case:

https://docs.xarray.dev/en/latest/user-guide/io.html#reading-encoded-data

(They say data is "always lazy loaded from netcdf", although the docs for
`load_dataset` say the opposite.)

*If* `load_dataset` is putting all data in memory, and `open_dataset` is
lazy-loading data, this should drastically reduce the amount of memory needed
to add footprints to the object store.

* **Please check if the PR fulfills these requirements**

- [x] Closes #749 (partially... I suggested using a script to compress anything already in the object store, but maybe it won't matter too much)
- [x] Closes #454 
- [x] [Tests added and passed](https://docs.openghg.org/development/python_devel.html#testing) if fixing a bug or adding a new feature
- [x] All code checks passing - `black --line-length 110` run over code, `mypy` and `flake8` not showing any errors
- [x] Documentation and tutorials updated/added
- [x] Added an entry in the latest `CHANGELOG.md` file if fixing a bug or adding a new feature
- [x] Added any new requirements to `requirements.txt` and `recipes/meta.yaml`

### CODE REVIEW

File: openghg/store/base/_datasource.py [hunk 1]
```diff
@@ -577,8 +579,60 @@ def save(self, bucket: str) -> None:
...
+                        if var not in do_not_compress:
+                            enc = {k: v for k, v in data[var].encoding.items() if k in valid_encoding_keys}
+                            enc.update(comp)
+                            encoding[var] = enc
 
+                    try:
+                        data.to_netcdf(filepath, engine="netcdf4", encoding=encoding)
+                    except RuntimeError:
+                        logger.warning(
+                            "Storing footprint for date range {daterange} without compression due to netCDF4 RuntimeError."
```
Comment (gareth-j) [inline openghg/store/base/_datasource.py:621]: This needs to be an f-string

Comment (gareth-j) [reply openghg/store/base/_datasource.py:621]: I'm getting some of these RuntimeErrors when I run the tests, can we have a look at what's causing them?

**Review by gareth-j — CHANGES_REQUESTED** [review]
Just a couple of questions and a quick fix for the f-string issue.

File: openghg/store/base/_datasource.py (hunk 1 above)
Comment (brendan-m-murphy) [reply openghg/store/base/_datasource.py]: I tried tracing what goes wrong. The variable that causes the problem is `rn_number_of_observations` for the WAO data in one of the fixtures in tests/analyse/conftest.py. This variable doesn't have chunking set... I'm not sure why it raises an issue. The error comes from the netcdf-c code base, and it's a pain to try to find where things are actually defined. For some reason the error comes from a function that has to do with chunking, but there is no chunking... so I have no idea what is wrong.

**Review by gareth-j — APPROVED** [review]
Looks good.

**Comment by rt17603** [comment]
@brendan-m-murphy - just to check - will these updates work ok with the current object store?

**Comment by brendan-m-murphy** [comment]
@rt17603 Yea I just branched off of devel. The lazy loading changing is just replacing `load_dataset` with `open_dataset`. The xarray docs say this should give you lazy loading. (I also replaced the binary IO part with just reading from a file, maybe that would be a problem for running in the cloud?) Actually the xarray docs say that they always use lazy loading, but the docs for `load_dataset` say it doesn't, while `open_dataset` does.

I haven't tested it by loading in a lot of footprints, but I could. I tried to write a test to check how much memory was being used by open vs. load dataset, but I couldn't get it to work (it's probably possible, but not straightforward because of how numpy stores things).

**Pull request merged**